ktor-client/ktor-client-core/common/src/io/ktor/client/engine/HttpClientEngine.kt

```diff
@@ -58,6 +58,43 @@ interface HttpClientEngine : CoroutineScope, Closeable {
             proceedWith(call)
         }
     }
+
+    /**
+     * Create call context and use it as a coroutine context to [execute] request.
+     */
+    private suspend fun executeWithinCallContext(requestData: HttpRequestData): HttpResponseData {
+        val callContext = createCallContext(requestData.executionContext)
+
+        return async(callContext + KtorCallContextElement(callContext[Job] as CompletableJob)) {
+            checkNotClosed()
+            execute(requestData)
+        }.await()
+    }
+
+    /**
+     * Checks that the client engine is not closed yet (works for [HttpClientEngineBase] extensions. If the client
+     * engine is already closed throws [ClientEngineClosedException].
+     */
+    private fun checkNotClosed() {
+        (this as? HttpClientEngineBase)?.also {
+            if (closed)
```

Comment:
Btw. Could you check if we get rid of the `closed` flag and use `CompletableJob` from context?

Reply:
Done.